sql/planner/optimizer.go

```diff
@@ -0,0 +1,431 @@
+			kv.V = precalculateExpr(kv.V)
+			if _, ok := kv.V.(expr.LiteralValue); !ok {
+				literalsOnly = false
+			}
+			t[i] = kv
+		}
+
+		// if literalsOnly is still true, it means we have a list of kvpairs
+		// that only contain constant values (ex: {"a": 1, "b": true}.
+		// We can transform that into a document.Document.
+		if literalsOnly {
+			var fb document.FieldBuffer
+			for i := range t {
+				fb.Add(t[i].K, document.Value(t[i].V.(expr.LiteralValue)))
+			}
+
+			return expr.LiteralValue(document.NewDocumentValue(&fb))
+		}
+	case expr.Operator:
+		// since expr.Operator is an interface,
+		// this optimization must only be applied to
+		// a few selected operators that we know about.
+		if !expr.IsAndOperator(t) &&
+			!expr.IsOrOperator(t) &&
+			!expr.IsArithmeticOperator(t) &&
+			!expr.IsComparisonOperator(t) {
+			return e
+		}
+
+		lh := precalculateExpr(t.LeftHand())
+		rh := precalculateExpr(t.RightHand())
+		t.SetLeftHandExpr(lh)
+		t.SetRightHandExpr(rh)
+
+		_, leftIsLit := lh.(expr.LiteralValue)
+		_, rightIsLit := rh.(expr.LiteralValue)
+		// if both operands are literals, we can precalculate them now
+		if leftIsLit && rightIsLit {
+			v, err := t.Eval(expr.EvalStack{})
+			// any error encountered here is unexpected
+			if err != nil {
+				panic(err)
+			}
+			// we replace this expression with the result of its evaluation
+			return expr.LiteralValue(v)
+		}
+	}
+
+	return e
+}
+
+// RemoveUnnecessarySelectionNodesRule removes any selection node whose
+// condition is a constant expression that evaluates to a truthy value.
+// if it evaluates to a falsy value, it considers that the tree
+// will not stream any document, so it returns an empty tree.
+func RemoveUnnecessarySelectionNodesRule(t *Tree) (*Tree, error) {
+	n := t.Root
+	var prev Node
+
+	for n != nil {
+		if n.Operation() == Selection {
+			sn := n.(*selectionNode)
+			if sn.cond != nil {
+				if lit, ok := sn.cond.(expr.LiteralValue); ok {
+					// if the expr is falsy, we return an empty tree
+					ok, err := document.Value(lit).IsTruthy()
+					if err != nil {
+						return nil, err
+					}
+					if !ok {
+						return &Tree{}, nil
+					}
+					// if the expr is truthy, we remove the node from the tree
+					if prev != nil {
+						prev.SetLeft(n.Left())
+					} else {
+						t.Root = n.Left()
+					}
+				}
+
+			}
+		}
+
+		prev = n
+		n = n.Left()
+	}
+
+	return t, nil
+}
+
+// UseIndexBasedOnSelectionNodeRule scans the tree for the first selection node whose condition is an
+// operator that satisfies the following criterias:
+// - implements the indexIteratorOperator interface
+// - one of its operands is a field selector that is indexed
+// - the other operand is a literal value or a parameter
+// If found, it will replace the input node by an indexInputNode using this index.
+func UseIndexBasedOnSelectionNodeRule(t *Tree) (*Tree, error) {
+	n := t.Root
+	var prev Node
+	var inputNode Node
+
+	// first we lookup for the input node
```

Comment:
Shouldn't the `Input` node be the first in the tree?

Reply:
It should be last because there can be multiple inputs (this is not yet supported) but only one output. If you take a look at the relational algebra tree examples, the inputs are always at the bottom, not at the root.